### ISSUE TITLE
cmd ip link set fail caused by dns_buffer

### ISSUE THREAD

**Issue by ChenYahui2019** [zebaz/xpress-dns#3]
```
# ip link set dev eth0 xdp obj xdp_dns_kern.o
ELF contains non-{map,call} related relo data in entry 1 pointing to section 65522! Compiler bug?!
Error fetching program/map!
```
I use centos8 and get fail above. The error was resloved after global dns_buffer was removed, but I think this is not the correct method.
```
# git diff
diff --git a/src/xdp_dns_kern.c b/src/xdp_dns_kern.c
index c91ed57..17f08bd 100644
--- a/src/xdp_dns_kern.c
+++ b/src/xdp_dns_kern.c
@@ -56,11 +56,11 @@ static inline void modify_dns_header_response(struct dns_hdr *dns_hdr);
 static inline void update_ip_checksum(void *data, int len, uint16_t *checksum_location);
 static inline void copy_to_pkt_buf(struct xdp_md *ctx, void *dst, void *src, size_t n);
 static inline void swap_mac(uint8_t *src_mac, uint8_t *dst_mac);
-char dns_buffer[512];

 SEC("prog")
 int xdp_dns(struct xdp_md *ctx)
 {
+char dns_buffer[64];
     void *data_end = (void *)(unsigned long)ctx->data_end;
     void *data = (void *)(unsigned long)ctx->data;
```

**Comment by zebaz** [comment]
I've seen similar problems with BCC, as well as with older clang versions. Can you share the compiler version you are using?
In any case, I will look into an appropriate solution to see if a global dns_buffer could be avoided altogether.

The workaround you've provided is fine, but note that the dns_buffer will be put unto the BPF stack, which is only 512 bytes long. When further extending the program you will run out of stack space fast for large dns_buffer sizes.

**Comment by ChenYahui2019** [comment]
```
# uname -r
4.18.0-193.el8.jd_016.x86_64
# clang --version
clang version 10.0.1 (Red Hat 10.0.1-1.module_el8.3.0+467+cb298d5b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
# llc --version
LLVM (http://llvm.org/):
  LLVM version 10.0.1
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: broadwell

  Registered Targets:
    aarch64    - AArch64 (little endian)
    aarch64_32 - AArch64 (little endian ILP32)
    aarch64_be - AArch64 (big endian)
    amdgcn     - AMD GCN GPUs
    arm        - ARM
    arm64      - ARM64 (little endian)
    arm64_32   - ARM64 (little endian ILP32)
    armeb      - ARM (big endian)
    bpf        - BPF (host endian)
    bpfeb      - BPF (big endian)
    bpfel      - BPF (little endian)
    mips       - MIPS (32-bit big endian)
    mips64     - MIPS (64-bit big endian)
    mips64el   - MIPS (64-bit little endian)
    mipsel     - MIPS (32-bit little endian)
    nvptx      - NVIDIA PTX 32-bit
    nvptx64    - NVIDIA PTX 64-bit
    ppc32      - PowerPC 32
    ppc64      - PowerPC 64
    ppc64le    - PowerPC 64 LE
    r600       - AMD GPUs HD2XXX-HD6XXX
    systemz    - SystemZ
    thumb      - Thumb
    thumbeb    - Thumb (big endian)
    x86        - 32-bit X86: Pentium-Pro and above
    x86-64     - 64-bit X86: EM64T and AMD64

```
I think using a map not global variable could resolve this issue.

**Comment by ChenYahui2019** [comment]
I have another issue: `bpf_xdp_adjust_tail` can't extend packet in centos8. 
Is there any way to resolve this problem?

**Comment by zebaz** [comment]
Support for global variables in BPF has been merged in kernel version 5.2
Extending packet size with bpf_xdp_adjust_tail has been implemented in kernel 5.8, see https://github.com/zebaz/xpress-dns#requirements

Seeing that you’re using kernel 4.18, the only solution for the program to work is to upgrade to kernel 5.8+

**Comment by ChenYahui2019** [comment]
OK, I get it.
Thx a lot
